javascript/ql/src/semmle/javascript/frameworks/Typeahead.qll

```diff
@@ -52,31 +52,29 @@ module Typeahead {
 
     override DataFlow::Node getADataNode() { none() }
 
+    DataFlow::SourceNode ref(DataFlow::TypeTracker t) {
+      t.start() and
+      result instanceof RemoteBloodhoundClientRequest
+      or
+      exists(DataFlow::TypeTracker t2 | result = ref(t2).track(t2, t))
+    }
+
+    DataFlow::SourceNode ref() { result = ref(DataFlow::TypeTracker::end()) }
+
     override DataFlow::Node getAResponseDataNode(string responseType, boolean promise) {
       responseType = "json" and
       promise = false and
-      /*
-       * exists(TypeAheadSuggestionTaintStep step |
-       *          (
-       *            this.flowsTo(step)
-       *            or
-       *            this.getAMethodCall("ttAdapter").flowsTo(step)
-       *          ) and
-       *          step.step(_, result)
-       *        )
-       *        or
-       */
-
-      // the first occurrence of the responseDataNode can be very disconnected from the instantiation of Bloodhound
-      // So I do this trick to get a taint-path that is readable to a developer.
-      // The above (possibly with added type-tracking) would be the correct way, but which gives unhelpful feedback to developers.
-      result = this
+      exists(TypeaheadSuggestionTaintStep step |
+        ref() = step.getALocalSource() or ref().getAMethodCall("ttAdapter") = step
+      |
+        result = step.getSuccessor()
+      )
     }
   }
 
   /**
-   * A function that generates suggestions to typeahead. 
-   * Matches `$(...).typeahead(..., { templates: { suggestion: <this> } })`. 
+   * A function that generates suggestions to typeahead.js.
+   * Matches `$(...).typeahead({..}, { templates: { suggestion: <this> } })`.
```

Comment:
Lets move the implementation detail of the pattern we happen to recognize out of the public docstring and into a line comment in the charpred.